File: 0008 字符串转换整数 (atoi)/main.py
class Solution:
    def myAtoi(self, s: str) -> int:
        num = 0
        a = False
        Pos = True
        for i in range(len(s)):
            c = s[i]
            if(c != " "):
                if(ord("0") <= ord(c) <= ord("9")):
                    if(not a):
                        Pos = False if (i != 0 and s[i-1] == "-") else True
                    num *= 10
                    num += int(c)
                    a = True
                else:
                    if(not a):
                        if(c == "+"):
                            Pos = True
                            a=True
                        elif(c == "-"):
                            Pos = False
                            a=True
                        else:
                            num = num if Pos else -num
                            if(Pos):
                                return 2**31-1 if num > 2**31-1 else num
                            else:
                                return -2**31 if num < -2**31 else num
                    else:
                        num = num if Pos else -num
                        if(Pos):
                            return 2**31-1 if num > 2**31-1 else num
                        else:
                            return -2**31 if num < -2**31 else num
            else:
                if(a):
                    num = num if Pos else -num
                    if(Pos):
                        return 2**31-1 if num > 2**31-1 else num
                    else:
                        return -2**31 if num < -2**31 else num
        num = num if Pos else -num
        if(Pos):
            return 2**31-1 if num > 2**31-1 else num
        else:
            return -2**31 if num < -2**31 else num
        


a = Solution()
print(a.myAtoi("982"))

File: 0008 字符串转换整数 (atoi)/main.cpp
class Solution {
public:
    int myAtoi(string s) {
        int res=0;
        int i=0;
        while(s[i] == ' ')
            i++;
        int c=0;
        if(s[i] >= '0' && s[i]<='9' ||s[i]=='+'){
            if(s[i]=='+')
                i++;
            for(;i<s.size();i++){
                c=s[i]-'0';
                if(c>=0 && c<=9){
                    if( res > INT_MAX/10 || (res == INT_MAX/10 && c>=8)){
                        return INT_MAX;
                    }
                    res = res*10 + c;
                }else{
                    return res;
                }
            }
        }else if(s[i]=='-'){
            for(i++;i<s.size();i++){
                c=s[i]-'0';
                if(c>=0 && c<=9){
                    if( res < INT_MIN/10 || (res == INT_MIN/10 && c>=9)){
                        return INT_MIN;
                    }
                    res = res*10 - c;
                }else{
                    return res;
                }
            }
        }
        return res;
    }
};